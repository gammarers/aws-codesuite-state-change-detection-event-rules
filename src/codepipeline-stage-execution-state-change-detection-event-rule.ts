import { InvalidInternalDefinitionParameterError } from '@gammarers/aws-cdk-errors';
import * as events from 'aws-cdk-lib/aws-events';
import { Construct } from 'constructs';

/**
 * @TODO: Not yet supported Omit
 * https://github.com/aws/jsii/issues/4468
 * type omitKeys = 'eventPattern';
 * export interface CodePipelineStateChangeDetectionEventRuleProps extends Omit<events.RuleProps, 'eventPattern'> {}
 */

export enum CodePipelineStageExecutionState {
  CANCELED = 'CANCELED',
  FAILED = 'FAILED',
  RESUMED = 'RESUMED',
  STARTED = 'STARTED',
  STOPPED = 'STOPPED',
  STOPPING = 'STOPPING',
  SUCCEEDED = 'SUCCEEDED',
}

export interface CodePipelineStageExecutionStateChangeDetectionEventRuleProps extends events.RuleProps {
  readonly targetStates?: CodePipelineStageExecutionState[];
}

export class CodePipelineStageExecutionStateChangeDetectionEventRule extends events.Rule {
  constructor(scope: Construct, id: string, props: CodePipelineStageExecutionStateChangeDetectionEventRuleProps) {
    super(scope, id, {
      ...props,
      eventPattern: ((): events.EventPattern => {
        if (props.eventPattern) {
          throw new InvalidInternalDefinitionParameterError('eventPattern');
        }
        return {
          source: ['aws.codepipeline'],
          detailType: ['CodePipeline Stage Execution State Change'],
          //          detail: {
          //            state: props.targetStates,
          //          },
          detail: (() => {
            if (props.targetStates) {
              return {
                state: props.targetStates,
              };
            }
            return undefined;
          })(),
        };
      })(),
    });
  }
}