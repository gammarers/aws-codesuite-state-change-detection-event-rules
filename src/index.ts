export * from './codepipeline-execution-state-change-detection-event-rule';
export * from './codepipeline-stage-execution-state-change-detection-event-rule';