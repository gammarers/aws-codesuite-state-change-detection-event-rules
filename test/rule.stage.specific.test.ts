import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import * as events from 'aws-cdk-lib/aws-events';
import { CodePipelineExecutionState, CodePipelineExecutionStateChangeDetectionEventRule } from '../src';

describe('Default Rule Check', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const rule = new CodePipelineExecutionStateChangeDetectionEventRule(stack, 'CodePipelineExecutionStateChangeDetectionEventRule', {
    ruleName: 'codepipeline-state-change-detection-event-rule',
    targetStates: [
      CodePipelineExecutionState.FAILED,
      CodePipelineExecutionState.CANCELED,
    ],
  });

  it('Is Rule', async () => {
    expect(rule).toBeInstanceOf(events.Rule);
  });

  const template = Template.fromStack(stack);

  it('Should match event rule.', async () => {
    template.hasResourceProperties('AWS::Events::Rule', Match.objectEquals({
      Name: 'codepipeline-state-change-detection-event-rule',
      State: 'ENABLED',
      EventPattern: Match.objectEquals({
        'source': [
          'aws.codepipeline',
        ],
        'detail-type': [
          'CodePipeline Pipeline Execution State Change',
        ],
        'detail': {
          state: Match.arrayEquals(['FAILED', 'CANCELED']),
        },
      }),
    }));
  });

  it('Should match snapshot.', async () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
