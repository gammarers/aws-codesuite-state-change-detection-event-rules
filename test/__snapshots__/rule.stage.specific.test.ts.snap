// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Rule Check Should match snapshot. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CodePipelineExecutionStateChangeDetectionEventRule98882579": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "state": [
              "FAILED",
              "CANCELED",
            ],
          },
          "detail-type": [
            "CodePipeline Pipeline Execution State Change",
          ],
          "source": [
            "aws.codepipeline",
          ],
        },
        "Name": "codepipeline-state-change-detection-event-rule",
        "State": "ENABLED",
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
